symbolic-base/src/ZkFold/Symbolic/Data/UInt.hs

```diff
@@ -265,33 +273,48 @@ instance
                         newI <- newAssigned (\j -> j xI + scale ((2 :: Natural) ^ xN) (j yI))
                         helper ((newN, newI) : ys) acc
 
-instance
-    ( Symbolic c
-    , KnownNat n
-    , KnownNat r
-    , KnownRegisterSize rs
-    , r ~ NumberOfRegisters (BaseField c) n rs
-    , KnownNat (Ceil (GetRegisterSize (BaseField c) n rs) OrdWord)
-    , NFData (c (Vector r))
-    ) => SemiEuclidean (UInt n rs c) where
-
-    divMod numerator d = bool @(Bool c) (q, r) (zero, zero) (d == zero)
-        where
-            (q, r) = Haskell.foldl longDivisionStep (zero, zero) [value @n -! 1, value @n -! 2 .. 0]
-
-            numeratorBits :: ByteString n c
-            numeratorBits = from numerator
-
-            addBit :: UInt n rs c -> Natural -> UInt n rs c
-            addBit ui bs = ui + bool @(Bool c) zero one (isSet numeratorBits bs)
-
-            longDivisionStep
-                :: (UInt n rs c, UInt n rs c)
-                -> Natural
-                -> (UInt n rs c, UInt n rs c)
-            longDivisionStep (q', r') i =
-                let rs = force $ addBit (r' + r') (value @n -! i -! 1)
-                 in bool @(Bool c) (q', rs) (q' + fromConstant ((2 :: Natural) ^ i), rs - d) (rs >= d)
+instance ( Symbolic c, KnownNat n, KnownRegisterSize r
+         , KnownRegisters c n r
+         , regSize ~ GetRegisterSize (BaseField c) n r
+         , KnownNat (Ceil regSize OrdWord)
+         ) => SemiEuclidean (UInt n r c) where
+    divMod num@(UInt nm) den@(UInt dn) =
```

Comment:
Can you please add a few comments about what each function inside `where` clause does and how everything works in general? I get the general idea but not the details

Reply:
Added!